### PR TITLE
Issue#915,and fix the issue that can not render correctly when change…

### PR DESCRIPTION
**Pull request by dblleaf**
… carousel DotPosition

[[中文版模板 / Chinese template](https://github.com/ant-design-blazor/ant-design-blazor/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
Fix issue #915 

### 💡 Background and solution
1. #915 
2. The carousel [demo](https://ant-design-blazor.gitee.io/en-US/components/carousel#components-carousel-demo-Position) can't render normally after setting Left of Right position.
image

### 📝 Changelog
1. Fix issue #915 and so on.
2. Refator the Carousel.

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  Fix issue #915 and so on;  Refator the Carousel.  |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by ElderJames** [comment]
Thank you for contribution @dblleaf .

When I was running the Carousel page in Server side, clicking on the browser refresh threw a lot of exceptions. Could you please fix it?

当我在 Server Side 模式运行时 Carousel页面时，点击浏览器的刷新会抛出很多异常。请问可以解决一下吗？

File: components/carousel/Carousel.razor.cs [hunk 1]
```diff
@@ -112,120 +127,70 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
...
         internal void RemoveSlick(CarouselSlick slick)
         {
-            _slicks.Remove(slick);
+            var slicks = new List<CarouselSlick>(_slicks ?? new List<CarouselSlick>());
+            slicks.Remove(slick);
+            _slicks = slicks;
+            InvokeAsync(async () =>
+            {
+                await InvokeAsync(StateHasChanged).ConfigureAwait(false);
+            });
```
Comment (ElderJames) [inline components/carousel/Carousel.razor.cs:149]:
```suggestion
            _slicks ??= new List<CarouselSlick>();
            _slicks.Remove(slick);
```

File: components/carousel/Carousel.razor.cs [hunk 2]
```diff
@@ -112,120 +127,70 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
...
-            if (_activeSlick == null)
+            var slicks = new List<CarouselSlick>(_slicks ?? new List<CarouselSlick>());
+            slicks.Add(slick);
+            _slicks = slicks;
+            if (ActiveSlick == null)
             {
-                Activate(0);
+                Activate(_slicks[0]);
             }
+            InvokeStateHasChanged();
```
Comment (ElderJames) [inline components/carousel/Carousel.razor.cs:161]:
```suggestion
            _slicks ??= new List<CarouselSlick>();
            _slicks.Add(slick);
            if (ActiveSlick == null)
            {
                Activate(_slicks[0]);
            }
            InvokeStateHasChanged();
```

**Comment by ElderJames** [comment]
这个PR修改后，渐显demo失效

文档上的是好的
image

**Comment by dblleaf** [comment]
现在这些明面上的bug修好了，但是还有隐藏的，比如Slick是混合添加进去的
```razor
<div>
    <Carousel>
        <CarouselSlick>1</CarouselSlick>
        <CarouselSlick>2</CarouselSlick>
        @foreach (var item in list)
        {
            <CarouselSlick>
                <h3>@item</h3>
            </CarouselSlick>
        }
        <CarouselSlick>3</CarouselSlick>
    </Carousel>
    <Button OnClick="Add">+</Button>
    <Button OnClick="Remove">-</Button>
</div>
@code{
    List<int> list = new List<int>() { 1, 2, 3, 4 };
    void Add()
    {
        var newList = new List<int>(list);
        newList.Add(list.Count + 1);
        this.list = newList;
    }
    void Remove()
    {
        var newList = new List<int>(list);
        newList.RemoveAt(list.Count - 1);
        this.list = newList;
    }
}
```
那么Slick的Index是按照添加进去的顺来的，而不是stick-list子元素的顺序，虽然表面看着没有问题，但是slick-active设置的目标好像是有问题的，我翻遍的blazor的文档和一些demo并没有找到比较何是的解决办法。

**Comment by ElderJames** [comment]
这种情况应该可以加一个Key属性来控制

File: components/carousel/Carousel.razor.cs (hunk 1 above)
Comment (ElderJames) [reply components/carousel/Carousel.razor.cs:157]: 这里可以不调用 StateHasChanged 吗？

Comment (dblleaf) [reply components/carousel/Carousel.razor.cs:157]: 正如我下面那个demo，如果不调用，删除数据但是渲染出的元素还在

**Review by ElderJames — APPROVED** [review]
It help a lot. Thanks!

**Pull request merged**